### PR TITLE
test(ec2): add security related unit tests

### PR DESCRIPTION
**Pull request by Hweinstock**
## Problem
We want unit tests to ensure certain security measures are maintained in the code. Properties tested in this PR include:
1. create inline policy with minimal actions necessary to perform desired action
2. prompt user to confirm attached role and allow cancellation. 
3. test run the ssh agent before using it for remote connections. 

## Solution
For each test we do the following:

1. assert that the actions we are adding to the policy remains as the core 6. 
2. use `getTestWindow()` to ensure user's are able to see policies added, and cancel the process. 
3. spy on `tryRun` to ensure it is used before returning the ssh path. 

As part of (3), we refactor the pathfinding code to be contained in its own class. This makes the code easier to test since we can now use `sinon.spy` and `sinon.stub`. It also allows us to move some of the shared logic into one place. 

---

License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by hayemaxi — APPROVED** [review]
lgtm!

File: packages/core/src/shared/utilities/pathFinder.ts
Comment (justinmk3) [inline packages/core/src/shared/utilities/pathFinder.ts:1]: Please avoid renaming this file. And the introduction of the "class" abstraction looks unnecessary, or at least it should wait until all feature-branches are merged in the upcoming weeks. Because there are similar changes on a feature branch (but without the class abstraction), so if your necessary changes are small, it will help avoid big noisy merge conflicts in this file.

Comment (Hweinstock) [reply packages/core/src/shared/utilities/pathFinder.ts:1]: I introduced the class to make it easier to spy/stub the methods. I wanted to have tests that assert some of these functions were called. Is there a natural way to do this without the class abstraction?

Comment (justinmk3) [reply packages/core/src/shared/utilities/pathFinder.ts:1]: Maybe not, but the test added here only stubs `findSshPath` and the test seems over-specific. Is it important to test the "dry run" behavior? Testing the actual result is usually much stronger coverage. Adding a test for `findSshPath` is worth doing.

Meanwhile, avoiding churn in pathFind.ts will be very helpful for upcoming feature-branch merges.

Comment (Hweinstock) [reply packages/core/src/shared/utilities/pathFinder.ts:1]: I agree that its a little over-specific to the implementation, but AppSec was asking for a test to show we perform some kind of "validation" of the ssh path found and I am unsure how else to show that. 

Testing `findSshPath` without stubs would involve some manipulation of the test environment $PATH variable, which seems risky unless we already have some well-tested utility code to do this for us. 

I found another way to stub/spy the method that significantly reduces churn in pathFind.ts. If we move `tryRun` to another file, and import it as a module, we can stub it in the tests. I think this also makes sense since `tryRun` doesn't fit in perfectly with all the path finding code. 

Still working on fixing the tests on Windows.

File: packages/core/src/shared/utilities/programUtils.ts
Comment (justinmk3) [inline packages/core/src/shared/utilities/programUtils.ts:18]: please leave it in pathFind.ts

it's not a goal for everything in shared/ to be named fooUtils. pathFind.ts was intentionally an attempt to avoid the "fooUtil" pattern where possible, to have more meaningful module names.

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/pathFind.test.ts:17]: i guess we could add this to `packages/core/src/shared/utilities/osUtils.ts` , since it's somewhat common.

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/pathFind.test.ts:66]: a way to test this without stubs would be to set $PATH to a test fixture directory.

File: packages/core/src/shared/utilities/programUtils.ts
Comment (justinmk3) [reply packages/core/src/shared/utilities/programUtils.ts:18]: oh, i just saw this:

> I found another way to stub/spy the method that significantly reduces churn in pathFind.ts. If we move `tryRun` to another file, and import it as a module, we can stub it in the tests.

in that case, let's name the new module pathFind2.ts. no point in pretending that it is something else.

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (Hweinstock) [reply packages/core/src/test/shared/utilities/pathFind.test.ts]: I found a function in `src/shared/vscode/env.ts` that actually did the same thing, so I removed it.

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (Hweinstock) [reply packages/core/src/test/shared/utilities/pathFind.test.ts]: I see. I was nervous about having code messing with $PATH in case something went wrong. Tried to safely wrap it in a reusable util function and added some tests for it. I think with this utility, the tests feel much more useful, so thanks for pushing me in this direction.

File: packages/core/src/shared/utilities/pathFind.ts
Comment (justinmk3) [inline packages/core/src/shared/utilities/pathFind.ts:29]: can the function stay in this location? that will help avoid merge conflicts with incoming feature-branches .

File: packages/core/src/test/testUtil.ts
Comment (justinmk3) [inline packages/core/src/test/testUtil.ts:630]: not a blocker, but future idea: this could be generalized to `withEnv()` or even just `with()`:

```
with({
  env: ...
})
```

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/pathFind.test.ts:50]: if we just set $PATH here directly, can `afterEach` restore it? 

the "with" pattern can be useful but it might be confusing if we have different ways of saving/storing state in tests (but maybe there is a good reason for it)

Comment (Hweinstock) [reply packages/core/src/test/shared/utilities/pathFind.test.ts]: Yeah, this definitely reduces the amount of code as well. My only concern is that adding more behavior into `beforeEach` and `afterEach` can make it hard to isolate the behavior of each test, and sometimes makes it harder to add tests if it doesn't fit with the pattern. 

I think this is still the better route since it is inline with the rest of the test code (ex. sinon, clock).

File: packages/core/src/test/shared/utilities/testUtils.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/testUtils.test.ts:19]:
```suggestion
describe('setEnv', function () {
```

File: packages/core/src/test/testUtil.ts
Comment (justinmk3) [inline packages/core/src/test/testUtil.ts:623]: this should return a shallow copy:
```suggestion
    return { ...process.env }
```

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/pathFind.test.ts:53]: if `readEnv` returns a shallow copy, then this becomes much more conventional/familiar:
```suggestion
            process.env = previousEnv
```

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/pathFind.test.ts:64]: if `readEnv` returns a shallow copy, then this becomes much more conventional/familiar:

```suggestion
            process.env.PATH = workspace.uri.fsPath
```

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/pathFind.test.ts:128]: Nice! This gives a lot more confidence that spies.

**Review by justinmk3 — APPROVED** [review]

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/pathFind.test.ts:102]: since we are now shallow-copying `env` in beforeEach, we can simply directly assign `env.PATH` here. `envWithNewPath` was only needed when beforeEach was not shallow-copied.
```suggestion
            process.env.PATH = workspace.uri.fsPath
```

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/pathFind.test.ts:64]:
```suggestion
            process.env.PATH = workspace.uri.fsPath
```

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/pathFind.test.ts:83]:
```suggestion
            process.env.PATH = workspace.uri.fsPath
```

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/pathFind.test.ts:49]: if we aren't going to store the whole env, we can just save PATH directly.
```suggestion
            previousPath = testutil.env.PATH
```

File: packages/core/src/test/shared/utilities/pathFind.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/pathFind.test.ts:46]:
```suggestion
        let previousPath: string | undefined
```

File: packages/core/src/test/shared/utilities/testUtils.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/testUtils.test.ts:12]:
```suggestion
describe('copyEnv', function () {
```

**Review by justinmk3 — APPROVED** [review]

File: packages/core/src/test/shared/utilities/testUtils.test.ts
Comment (justinmk3) [inline packages/core/src/test/shared/utilities/testUtils.test.ts:16]: PATH should not be undefined on Windows. Even if it did, if I'm reading the test correctly, it looks like it should still pass

**Pull request merged**
